include/ginkgo/core/base/dim.hpp

```diff
@@ -152,6 +152,20 @@ struct dim {
         return dim(x.first_ * y.first_, x.rest_ * y.rest_);
     }
 
+    /**
+     * A stream operator overload for dim
+     *
+     * @param os  stream object
+     * @param x  dim object
+     *
+     * @return a stream object appended with the dim output
+     */
+    friend GKO_ATTRIBUTES std::ostream &operator<<(std::ostream &os,
+                                                   const dim &x)
+    {
+        return os << "(" << x.first_ << " , " << x.rest_ << ")";
```

Comment:
That will lead to a gang of owls `(0,(0,(0,0)))` for higher dimensionalities. I guess that might not be what we want?

Reply:
I am not sure how else to accurately represent this. I guess our main use case would be `dim<2>`, for which this output should be okay.

Reply:
I would suggest this:
Unfortunately, I need the `friend` declaration with the `dim<N+1>`, so I can call `rest_.print_to(os)`. It does seem to work properly, but it is more code than your suggestion.
```
#include <iostream>

template <int N>
class dim {
    static_assert(N > 1, "Internal Assert!");
    friend class dim<N + 1>;

   public:
    template <typename... Args>
    dim(int first, Args &&... args)
        : current_{first}, rest_(std::forward<Args>(args)...) {}

    friend std::ostream &operator<<(std::ostream &os, const dim &x) {
        os << '(';
        x.print_to(os);
        os << ')';
        return os;
    }

   private:
    void inline print_to(std::ostream &os) const {
        os << current_ << ", ";
        rest_.print_to(os);
    }

    int current_;
    dim<N - 1> rest_;
};

template <>
class dim<1> {
    friend class dim<2>;

   public:
    dim(int first) : current_{first} {}

    friend std::ostream &operator<<(std::ostream &os, const dim &x) {
        os << '(';
        x.print_to(os);
        os << ')';
        return os;
    }

   private:
    void inline print_to(std::ostream &os) const { os << current_; }

    int current_;
};

template <>
class dim<0> {
   public:
    dim() {}

    friend std::ostream &operator<<(std::ostream &os, const dim &x) {
        os << '(';
        os << ')';
        return os;
    }

   private:
    void inline print_to(std::ostream &os) const {}
};

int main() {
    dim<3> test3(1, 2, 3);
    dim<2> test2(5, 6);
    dim<1> test1(7);
    dim<0> test0;

    std::cout << test3 << '\n'
              << test2 << '\n'
              << test1 << '\n'
              << test0 << '\n';
}
```

Reply:
Thanks! That works quite nicely.